### PR TITLE
Fix PIMD: bugfix in spring energy, and adding virial estimator

### PR DESCRIPTION
**Pull request by yotamfe**
**Summary**
Hi,

Two proposed changes to the quantities that Fix PIMD exposes for each step:

1. Fix a bug in the calculation of the spring energy, using the position difference rather than the position, and including the spring constant factor.
2. Provide also the current value of the virial, which is useful as an estimator for the kinetic energy of the original quantum system. It is necessary to add this as an additional quantity because the estimator cannot otherwise be recovered from the already available *total* force.

**Related Issue(s)**

None.

**Author(s)**

Barak Hirshberg
Sagi Meir

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

No issues expected.

**Implementation Notes**

1. We tested the fix for the spring energy by comparing the reported values to values we computed post-processing from the positions, in a simulation of several particles in a harmonic trap.
2. We tested the virial by estimating the energy of several particles in a harmonic trap and comparing the resulting total energy to the analytical solution for non-interacting quantum harmonic oscillators.

**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

### CODE REVIEW

**Review by athomps — APPROVED** [review]
I approve.

**Comment by akohlmey** [comment]
@yotamfe thanks for your submission. It looks like there should be some documentation added for the additional property that can be read from the fix (and specifically it should state that it is a scalar and not the more typical 3x3 tensor). 

This would normally go to a section titled:
`Restart, fix_modify, output, run start/stop, minimize info`

Which seems to be missing altogether. So we should take this opportunity to add one and fill in the respective information. You can compare to other fixes what is added there.

Thanks in advance.

**Comment by yotamfe** [comment]
> 

@akohlmey Thanks for reviewing this! I'll look at the other fixes and add the relevant documentation.

**Comment by akohlmey** [comment]
> > 
> 
> @akohlmey Thanks for reviewing this! I'll look at the other fixes and add the relevant documentation.

Great. When you are done and have updated your branch, please click on the "Ready for review" button, so we know you are done. Thanks.

**Comment by BarakHirshberg** [comment]
Hi Axel and Yotam,
Just found two small bugs in the testing of fmass and sp variables in the constructor of FixPIMD.
Added them to the same pull request, I hope that's OK.
Barak

**Comment by yotamfe** [comment]
> Hi Axel and Yotam, Just found two small bugs in the testing of fmass and sp variables in the constructor of FixPIMD. Added them to the same pull request, I hope that's OK. Barak

The fix for fmass is consistent with documentation that says that the parameter can be set to the number of quasi-beads P (<code>np</code> in the code).

**Comment by yotamfe** [comment]
@akohlmey I added the documentation. There's one point I'm not sure about: From what I gather from the code, it seems that the binary restart files contain only the state of the Nose-Hoover thermostat, and this is what I wrote in the documentation. (We didn't write this feature, nor have experience with it.) Does it make sense that the x,y,z positions are not stored in that file?

**Comment by akohlmey** [comment]
> it seems that the binary restart files contain only the state of the Nose-Hoover thermostat

A fix only needs to store information in a restart that it "owns". This is how restarting is orgnized in LAMMPS. That is, position data will be stored by the atom style, potential data by the pair style and so on. Because of that, each command documents which of its properties it saves in a restart file or not. That information is needed so that can know whether a) a clean restart is possible at all and b) which commands need to be included in the input when restarting (e.g. whether the pair_style and pair_coeff commands need to be repeated). For fixes generally the command needs to be repeated with the same fix ID.

File: doc/src/fix_pimd.rst [hunk 1]
```diff
@@ -149,6 +149,31 @@ related tasks for each of the partitions, e.g.
    restart 1000 system_${ibead}.restart1 system_${ibead}.restart2
    read_restart system_${ibead}.restart2
 
+Restart, fix_modify, output, run start/stop, minimize info
+"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
+
+This fix writes the state of the Nose/Hoover thermostat over all quasi-beads to
```
Comment (BarakHirshberg) [inline doc/src/fix_pimd.rst]: Thanks, @akohlmey, for explaining how the restart files work on LAMMPS.
In that context, do you think this phrasing is adequate?
Is that how fixes usually describe the restart files? Just stating what they add to the binary?

Comment (akohlmey) [reply doc/src/fix_pimd.rst]: I will look at it later today. If needed, I will make some final changes myself.
I appreciate you both taking the time to add the missing text, even for the parts you didn't write. It is a big help.

**Pull request merged**